TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+#import <TWTValidation/TWTJSONSchemaParser.h>
+
+#import <TWTValidation/TWTJSONSchemaArrayASTNode.h>
+#import <TWTValidation/TWTJSONSchemaASTNode.h>
+#import <TWTValidation/TWTJSONSchemaBooleanValueASTNode.h>
+#import <TWTValidation/TWTJSONSchemaDepedencyASTNode.h>
+#import <TWTValidation/TWTJSONSchemaGenericASTNode.h>
+#import <TWTValidation/TWTJSONSchemaKeywordConstants.h>
+#import <TWTValidation/TWTJSONSchemaNamedPropertyASTNode.h>
+#import <TWTValidation/TWTJSONSchemaNumberASTNode.h>
+#import <TWTValidation/TWTJSONSchemaObjectASTNode.h>
+#import <TWTValidation/TWTJSONSchemaPatternPropertyASTNode.h>
+#import <TWTValidation/TWTJSONSchemaStringASTNode.h>
+#import <TWTValidation/TWTJSONSchemaTopLevelASTNode.h>
+#import <TWTValidation/TWTJSONSchemaValidTypesConstants.h>
+
+
+static NSString *const kTWTJSONExceptionErrorKey = @"error";
+NSString *const kTWTJSONSchemaParserErrorInfoDomainKey = @"TWTJSONSchemaParserErrorDomain";
+NSString *const kTWTJSONSchemaParserErrorInfoPathKey = @"path";
+NSString *const kTWTJSONSchemaParserErrorInfoInvalidObjectKey = @"object";
+
+
+@interface TWTJSONSchemaParser ()
+
+@property (nonatomic, strong) NSMutableArray *pathStack;
+
+@property (nonatomic, copy) NSDictionary *JSONSchema;
+
+@end
+
+
+@implementation TWTJSONSchemaParser
+
+- (instancetype)initWithJSONSchema:(NSDictionary *)topLevelSchema
+{
+    self = [super init];
+    if (self) {
+        _JSONSchema = [topLevelSchema copy];
+        _pathStack = [[NSMutableArray alloc] init];
+    }
+    return self;
+}
+
+
+- (TWTJSONSchemaTopLevelASTNode *)parseWithError:(NSError *__autoreleasing *)errorPointer
+{
+    [self checkObject:self.JSONSchema[kTWTJSONSchemaKeywordSchema] isOneOf:[NSSet setWithObject:kTWTJSONSchemaKeywordDraft4Path]];
+
+    TWTJSONSchemaTopLevelASTNode *topLevelNode = [[TWTJSONSchemaTopLevelASTNode alloc] init];
+    topLevelNode.schemaPath = self.JSONSchema[kTWTJSONSchemaKeywordSchema];
+
+    @try {
+        topLevelNode.schema = [self parseSchema:self.JSONSchema];
+    } @catch (NSException *exception) {
+        NSError *outError = exception.userInfo[kTWTJSONExceptionErrorKey];
+        if (errorPointer) {
+            *errorPointer = outError;
+            return nil;
+        }
+    }
+
+    return topLevelNode;
+}
+
+
+# pragma mark - Schema parser methods
+// starting point for all schema nodes (except top level)
+// directs the schema toward the appropriate parser method based on the value of "type"
+// type-specific parser methods are responsible for the entire process of creating a node from a schema, including:
+// 1) initializing the appropriate ASTNode 2) setting the type (if its not built into that ASTNode class) 3) parsing all general and type-specific properties 4) returning a completed node
```

Comment:
Put each item of this numbered list on its own line. Also use . instead of close parenthesis please.